File: src/truthound_dashboard/api/triggers.py
"""Trigger monitoring and webhook API endpoints.

This module provides endpoints for:
- Trigger monitoring status
- Webhook trigger reception
- Trigger status queries
"""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Header, HTTPException

from truthound_dashboard.core.scheduler import get_scheduler
from truthound_dashboard.schemas.triggers import (
    TriggerCheckStatus,
    TriggerMonitoringResponse,
    TriggerMonitoringStats,
    WebhookTestReceivedData,
    WebhookTestResponse,
    WebhookTriggerRequest,
    WebhookTriggerResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/triggers", tags=["triggers"])


@router.get(
    "/monitoring",
    response_model=TriggerMonitoringResponse,
    summary="Get trigger monitoring status",
)
async def get_trigger_monitoring() -> TriggerMonitoringResponse:
    """Get current trigger monitoring status.

    Returns:
        Trigger monitoring stats and schedule statuses.
    """
    scheduler = get_scheduler()
    status = scheduler.get_trigger_monitoring_status()
    schedules = await scheduler.get_trigger_check_statuses()

    # Calculate aggregate stats
    active_data_change = sum(
        1 for s in schedules if s["trigger_type"] == "data_change"
    )
    active_webhook = sum(
        1 for s in schedules if s["trigger_type"] == "webhook"
    )
    active_composite = sum(
        1 for s in schedules if s["trigger_type"] == "composite"
    )

    # Calculate average check interval
    check_intervals = [
        s.get("check_interval_minutes", 5) * 60 for s in schedules
    ]
    avg_interval = (
        sum(check_intervals) / len(check_intervals)
        if check_intervals
        else 300.0
    )

    # Find next scheduled check
    next_checks = [
        s["next_check_at"] for s in schedules if s.get("next_check_at")
    ]
    next_check = min(next_checks) if next_checks else None

    stats = TriggerMonitoringStats(
        total_schedules=len(schedules),
        active_data_change_triggers=active_data_change,
        active_webhook_triggers=active_webhook,
        active_composite_triggers=active_composite,
        total_checks_last_hour=status.get("checks_last_hour", 0),
        total_triggers_last_hour=status.get("triggers_last_hour", 0),
        average_check_interval_seconds=avg_interval,
        next_scheduled_check_at=next_check,
    )

    schedule_statuses = [
        TriggerCheckStatus(**s) for s in schedules
    ]

    return TriggerMonitoringResponse(
        stats=stats,
        schedules=schedule_statuses,
        checker_running=status.get("checker_running", False),
        checker_interval_seconds=status.get("checker_interval_seconds", 300),
        last_checker_run_at=status.get("last_checker_run_at"),
    )


@router.get(
    "/schedules/{schedule_id}/status",
    response_model=TriggerCheckStatus,
    summary="Get trigger status for a specific schedule",
)
async def get_schedule_trigger_status(schedule_id: str) -> TriggerCheckStatus:
    """Get trigger status for a specific schedule.

    Args:
        schedule_id: Schedule ID to query.

    Returns:
        Trigger status for the schedule.
    """
    scheduler = get_scheduler()
    schedules = await scheduler.get_trigger_check_statuses()

    for schedule in schedules:
        if schedule["schedule_id"] == schedule_id:
            return TriggerCheckStatus(**schedule)

    raise HTTPException(
        status_code=404,
        detail=f"Schedule {schedule_id} not found or not using monitored trigger type",
    )


@router.post(
    "/webhook",
    response_model=WebhookTriggerResponse,
    summary="Receive webhook trigger",
)
async def receive_webhook(
    request: WebhookTriggerRequest,
    x_webhook_signature: str | None = Header(
        default=None, description="HMAC-SHA256 signature for verification"
    ),
) -> WebhookTriggerResponse:
    """Receive and process an incoming webhook trigger.

    This endpoint is called by external systems (Airflow, Dagster, Prefect, etc.)
    to trigger validations when data pipelines complete.

    Args:
        request: Webhook trigger request.
        x_webhook_signature: Optional signature for verification.

    Returns:
        Webhook trigger response with triggered schedules.
    """
    scheduler = get_scheduler()

    result = await scheduler.trigger_webhook(
        source=request.source,
        event_type=request.event_type,
        payload=request.payload,
        schedule_id=request.schedule_id,
        source_id=request.source_id,
        signature=x_webhook_signature,
    )

    return WebhookTriggerResponse(
        accepted=result["accepted"],
        triggered_schedules=result["triggered_schedules"],
        message=result["message"],
        request_id=result["request_id"],
    )


@router.post(
    "/webhook/test",
    response_model=WebhookTestResponse,
    summary="Test webhook configuration",
)
async def test_webhook(
    source: str = "test",
    event_type: str = "test_event",
) -> WebhookTestResponse:
    """Test webhook endpoint without triggering any schedules.

    Useful for verifying connectivity and configuration.

    Args:
        source: Test source name.
        event_type: Test event type.

    Returns:
        Test result.
    """
    return WebhookTestResponse(
        message="Webhook endpoint is accessible",
        received=WebhookTestReceivedData(
            source=source,
            event_type=event_type,
        ),
    )

File: frontend/src/api/modules/triggers.ts
/**
 * Triggers API - Schedule triggers and webhook endpoints.
 */
import { request } from '../core'

// ============================================================================
// Types
// ============================================================================

export interface TriggerEvaluation {
  source_id: string
  triggered: boolean
  reason: string
}

export interface TriggerCheckStatus {
  schedule_id: string
  schedule_name: string
  source_id: string
  trigger_type: string
  is_active: boolean
  last_check_at: string | null
  next_check_at: string | null
  last_trigger_at: string | null
  last_result: TriggerEvaluation | null
  cooldown_remaining_seconds: number | null
  is_due_for_check: boolean
  check_count?: number
  trigger_count?: number
  priority?: number
  last_evaluation?: unknown
}

export interface TriggerMonitoringStats {
  total_schedules: number
  active_triggers: number
  triggers_by_type: Record<string, number>
  triggered_last_24h: number
  last_check_at: string | null
  active_data_change_triggers?: number
  active_webhook_triggers?: number
  active_composite_triggers?: number
  total_checks_last_hour?: number
  total_triggers_last_hour?: number
  next_scheduled_check_at?: string | null
}

export interface ScheduleStatus {
  id: string
  name: string
  source_id: string
  source_name: string
  trigger_type: string
  is_active: boolean
  last_run_at: string | null
  next_run_at: string | null
  cooldown_remaining_seconds: number | null
}

export interface TriggerMonitoringResponse {
  stats: TriggerMonitoringStats
  trigger_statuses: TriggerCheckStatus[]
  schedules?: ScheduleStatus[]
  checker_running?: boolean
  checker_interval_seconds?: number
  last_checker_run_at?: string | null
}

export interface WebhookTriggerRequest {
  source?: string
  event_type?: string
  source_ids?: string[]
  metadata?: Record<string, unknown>
}

export interface WebhookTriggerResponse {
  success: boolean
  message: string
  triggered_schedules: string[]
  validations_started: Array<{
    schedule_id: string
    validation_id: string
  }>
  source: string
  event_type: string
}

// ============================================================================
// API Functions
// ============================================================================

export async function getTriggerMonitoring(): Promise<TriggerMonitoringResponse> {
  return request<TriggerMonitoringResponse>('/triggers/monitoring')
}

export async function getScheduleTriggerStatus(
  scheduleId: string
): Promise<TriggerCheckStatus> {
  return request<TriggerCheckStatus>(`/triggers/schedule/${scheduleId}`)
}

export async function sendWebhookTrigger(
  data: WebhookTriggerRequest
): Promise<WebhookTriggerResponse> {
  return request<WebhookTriggerResponse>('/triggers/webhook', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function testWebhookEndpoint(
  source: string = 'test',
  eventType: string = 'test_event'
): Promise<{ success: boolean; message: string; received: { source: string; event_type: string } }> {
  return request('/triggers/webhook/test', {
    method: 'POST',
    params: { source, event_type: eventType },
  })
}
